libvips/foreign/openslideconnection.c

```diff
@@ -0,0 +1,252 @@
+	(at your option) any later version.
+
+	This program is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU Lesser General Public License for more details.
+
+	You should have received a copy of the GNU Lesser General Public License
+	along with this program; if not, write to the Free Software
+	Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+	02110-1301  USA
+
+ */
+
+/*
+#define VIPS_DEBUG
+ */
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif /*HAVE_CONFIG_H*/
+#include <glib/gi18n-lib.h>
+
+#ifdef HAVE_OPENSLIDE
+
+#include <vips/vips.h>
+#include <vips/internal.h>
+
+#include "pforeign.h"
+
+#include <openslide.h>
+
+// we keep this many connections around after they are closed, ready for reuse
+#define OPENSLIDECONNECTION (3)
+
+typedef struct _VipsOpenslideConnection {
+	// lock access to these vars here
+	GMutex lock;
+
+	int ref_count;
+	int time;
+	char *filename;
+	openslide_t *osr;
+} VipsOpenslideConnection;
+
+static GHashTable *vips_openslideconnection_cache = NULL;
+static GMutex vips_openslideconnection_lock;
+
+static void
+vips_openslideconnection_free(VipsOpenslideConnection *connection)
+{
+	VipsOpenslideConnection *cached =
+		g_hash_table_lookup(vips_openslideconnection_cache,
+			connection->filename);
+	g_assert(cached);
+	g_assert(cached == connection);
+	g_hash_table_remove(vips_openslideconnection_cache, connection->filename);
+
+	g_mutex_clear(&connection->lock);
+	VIPS_FREE(connection->filename);
+	VIPS_FREEF(openslide_close, connection->osr);
+	g_free(connection);
+}
+
+static void
+vips_openslideconnection_unused_cb(gpointer key, gpointer value, gpointer data)
+{
+	VipsOpenslideConnection *connection = (VipsOpenslideConnection *) value;
+	int *count = (int *) data;
+
+	g_mutex_lock(&connection->lock);
+
+	if (connection->ref_count == 0)
+		*count += 1;
+
+	g_mutex_unlock(&connection->lock);
+}
+
+static int
+vips_openslideconnection_unused(void)
+{
+	int unused;
+
+	unused = 0;
+	g_hash_table_foreach(vips_openslideconnection_cache,
+		vips_openslideconnection_unused_cb, &unused);
+
+	return unused;
+}
+
+static void
+vips_openslideconnection_oldest_cb(gpointer key, gpointer value, gpointer data)
+{
+	VipsOpenslideConnection *connection = (VipsOpenslideConnection *) value;
+	VipsOpenslideConnection **oldest = (VipsOpenslideConnection **) data;
+
+	g_mutex_lock(&connection->lock);
+
+	if (!*oldest ||
+		connection->time < (*oldest)->time)
+		*oldest = connection;
+
+	g_mutex_unlock(&connection->lock);
+}
+
+static VipsOpenslideConnection *
+vips_openslideconnection_oldest(void)
+{
+	VipsOpenslideConnection *oldest;
+
+	oldest = NULL;
+	g_hash_table_foreach(vips_openslideconnection_cache,
+		vips_openslideconnection_oldest_cb, &oldest);
+
+	return oldest;
+}
+
+static void
+vips_openslideconnection_trim(void)
```

Comment:
You could possibly simplify this path by replacing the `time` mechanism with a global LRU `GQueue` of unreferenced connections.

Reply:
You're right, I thought about using a queue rather than a timestamp, but we're only going to be keeping a couple of connections around (they are relatively "heavy", as you note), so I didn't think the extra complexity was worth it.

Reply:
Well, we're only keeping a couple idle connections, but potentially unlimited active connections depending on what libvips' caller does.

Reply:
I ended up implementing this in https://github.com/libvips/libvips/pull/4455.